api/maven-api-xml/src/main/java/org/apache/maven/api/xml/XmlNode.java

```diff
@@ -18,121 +18,244 @@
  */
 package org.apache.maven.api.xml;
 
+import javax.xml.stream.XMLStreamException;
+
+import java.io.Serializable;
+import java.io.StringWriter;
 import java.util.List;
+import java.util.ListIterator;
 import java.util.Map;
-
-import org.apache.maven.api.annotations.Experimental;
-import org.apache.maven.api.annotations.Immutable;
-import org.apache.maven.api.annotations.Nonnull;
-import org.apache.maven.api.annotations.Nullable;
-import org.apache.maven.api.annotations.ThreadSafe;
+import java.util.Objects;
+import java.util.function.Function;
 
 /**
- * An immutable xml node.
- *
- * @since 4.0.0
+ *  NOTE: remove all the util code in here when separated, this class should be pure data.
  */
-@Experimental
-@ThreadSafe
-@Immutable
-public interface XmlNode {
+public class XmlNode implements Serializable {
```

Comment:
Use an interface and a builder API, similar to requests / models in other parts of the API.

Reply:
This is exactly what I don't what to do. That just makes the API more complex than it needs to be.